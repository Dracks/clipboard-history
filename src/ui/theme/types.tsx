import { ButtonSize } from "@elastic/eui/src/components/button/button";

export interface ButtonProps {
    children?: any,
    onClick: ()=>void
    size?: ButtonSize
    disabled?: boolean
}

export interface CheckboxProps{
    id: string,
    value: boolean,
    text: string,
    onChange: (newValue: boolean)=>void
}

export interface SectionProps{
    title: string,
    children: any
}

export interface PageProps {
    children: any
}

export interface SelectProps {
    text: string,
    options: {[key:string]: string}
    selected: string
    onChange: (newValue: string)=>void
}

export interface TextFieldProps<T>{
    value: T,
    onChange?: (newValue:T)=>void
    disabled?: boolean
    onClick?: ()=>void
}