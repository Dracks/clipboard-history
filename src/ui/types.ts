export type Ipc = {
    on: (channel: string, cb:(...args: any[])=>void)=>void,
    removeAllListeners: (channel: string)=>void,
    send: (channel: string, data: any)=>void,
}