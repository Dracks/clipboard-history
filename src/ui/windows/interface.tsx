
export type WindowPageProps<T> = {
    data: T,
    save: (data:T)=>void
}