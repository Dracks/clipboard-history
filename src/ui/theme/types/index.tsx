
export interface CheckboxProps{
    id: string,
    value: boolean,
    text: string,
    onChange: (newValue: boolean)=>void
}

export interface SectionProps{
    title: string,
    children: any
}

export interface PageProps {
    children: any
}

export interface SelectProps {
    text: string,
    options: {[key:string]: string}
    selected: string
    onChange: (newValue: string)=>void
}

export interface TextFieldProps<T>{
    value: T,
    onChange: (newValue:T)=>void
}