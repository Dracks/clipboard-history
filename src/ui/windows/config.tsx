import React, { useCallback } from 'react';
import { Config, ShortcutsConfig } from "../../common/config";
import { ChangeContext } from '../../common/types';
import { mergeObjects } from '../../common/utils';
import { Checkbox, Page, Section, Spacer, TextField, useModal } from '../theme';
import { getGeneric } from '../utils/utils';
import { WindowPageProps } from './interface';
import SetShortcut from './set_shortcut';


const NOTIFICATIONS_EXPLANATIONS : {
    [key in ChangeContext]: string
} = {
    manual: "When some clipboard is selected from the tray",
    new: "When the clipboard changed",
    shortcut: "When the change is provoqued by a shortcut",
    start: "When the application starts"
}

const SHORTCUTS_EXPLANATIONS : {
    [key in keyof ShortcutsConfig]: string
} = {
    next: "The shortcut to select next value",
    previous: "The shortcut to select previous value",
    removeCurrent: "The shortcut to remove the current value from the history and move to next"
}

const Configuration = ({data, save}: WindowPageProps<Config>)=>{
    const config = data
    const setConfig = save
    const update = useCallback((newData: Partial<Config>)=>{
        setConfig(mergeObjects(config, newData))
    }, [config])
    const shortcuts = Object.keys(config.shortcuts) as any

    return <Page >
            <Section title="General" >
                <h2>History length</h2>
                <TextField.Number {...getGeneric(["historyLength"], config, update)}/>
                <Spacer />
                {shortcuts.map((key: keyof ShortcutsConfig)=>{
                    const {value, onChange} = getGeneric(["shortcuts", key], config, ()=>undefined)
                    const {modal, show} = useModal(({close})=><SetShortcut save={(shortcut)=>{onChange(shortcut);close()}}/>, key)
                    return <div key={key}>
                        {modal}
                        <div>{SHORTCUTS_EXPLANATIONS[key]}</div>
                        <TextField.String
                            value={value}
                            onClick={show}
                            disabled
                            />
                        <Spacer />
                    </div>
                })}
            </Section>
            <Spacer/>
            <Section title="notifications">
                {Object.values(ChangeContext).map((key:ChangeContext)=>(
                    <div key={key}>
                        <Checkbox
                            id={key}
                            text={NOTIFICATIONS_EXPLANATIONS[key]}
                            {...getGeneric(["notifications", key], config, update)}
                            />
                    </div>
                ))}
            </Section>
        </Page>
}

export default Configuration