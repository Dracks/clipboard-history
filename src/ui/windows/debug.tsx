import React from 'react';
import { Section } from '../theme';
import { WindowPageProps } from './interface';


const Debug=(args: WindowPageProps<String>)=>{
    return (
        <Section title="Section title!">
            {args.data}
        </Section>
    )
}

export default Debug