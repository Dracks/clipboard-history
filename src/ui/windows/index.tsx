import React, { useCallback } from "react";
import { PageData, PageDataSend, WindowPage } from "../../common/types";
import Configuration from "./config";
import Debug from "./debug";
import { WindowPageProps } from "./interface";

type PagesList = {
    [k in WindowPage]: React.ComponentType<WindowPageProps<PageData[k]>>
}

const PAGES_LIST : PagesList= {
    "Debug": Debug,
    "config": Configuration
}



const Content = ({data, setData}: {data: PageDataSend, setData:(a:PageDataSend)=>void})=>{
    const { name } = data;
    const save = useCallback((data)=>{
        setData({
            name,
            data
        })
    }, [name] )

    const Page = PAGES_LIST[name] as React.ComponentType<WindowPageProps<any>>
    return <Page data={data.data} save={save}/>
}

export default Content