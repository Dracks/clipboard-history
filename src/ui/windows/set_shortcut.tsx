import React, { useCallback, useContext } from 'react';
import { PlatformEnum } from '../../common/types';
import context from '../context';
import ShortcutsRegister from '../utils/shortcuts.register';

interface SetShortcutProps{
    save: (shortcut:string)=>void
}

const MAP_TO_STANDARD :{[k in PlatformEnum]: {[k:string]:string}} = {
    mac: {
        Meta: "Cmd"
    },
    linux: {},
    win: {
        Meta: "Win"
    }
}

const SetShortcut = ({save}: SetShortcutProps)=>{
    const {platform} = useContext(context)
    const mapFn = useCallback((k:string)=>{
        const mapper = MAP_TO_STANDARD[platform]
        return mapper[k] || k
    }, [platform])
    const saveFn = useCallback((keys:string[])=>{
        save(keys.map(mapFn).reduce((ac,e)=>ac+"+"+e))
    }, [mapFn, save])
    return <div>
        Pres keys to change it, when all where released, it will saved automatically
        <ShortcutsRegister onAllReleased={saveFn}>
        {({keys})=>{
            if (keys.length){
                return <div>Keys pressed: {keys.map(mapFn).reduce((ac,e)=>ac+", "+e)}</div>
            } else {
                return <div>Press some keys for start the new shortcut</div>
            }
        }}
        </ShortcutsRegister>
    </div>
}

export default SetShortcut