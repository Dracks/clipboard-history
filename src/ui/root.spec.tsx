import { act, render, RenderResult } from '@testing-library/react';
import React from 'react';
import { Observable, Subscriber } from "rxjs";
import { LoadPage, PlatformEnum } from "../common/types";
import Subject from './root';
import { WindowPageProps } from './windows/interface';

jest.mock('./windows/debug', ()=>{
    return ({save, data}: WindowPageProps<String>)=>(
        <div>
            {data}
            <button onClick={()=>save('other info')}>save</button>
        </div>
    )
})

const debugPage : LoadPage = {page: {data:"Debug data", name: "Debug"}, context: {platform: PlatformEnum.linux}}

describe('ui/root', ()=>{
    let onSave : jest.Mock
    let pageInfoSubscription: Subscriber<LoadPage>
    let titleMock: any
    let getElementsByTagNameSpied: jest.SpyInstance
    let query: RenderResult
    beforeEach(()=>{
        titleMock = {}
        getElementsByTagNameSpied = jest.spyOn(document, 'getElementsByTagName').mockReturnValue([titleMock] as any)
        onSave = jest.fn()
        const loadPage = Observable.create((s:Subscriber<LoadPage>)=>{
            pageInfoSubscription = s
        })
        query = render(<Subject onSave={onSave} loadPageInfo={loadPage}/>)
    })

    afterEach(()=>{
        getElementsByTagNameSpied.mockRestore()
    })

    it('Should render basically', ()=>{
        expect(titleMock).toEqual({innerHTML: 'Loading...'})
    })

    it('Should render the debug screen', ()=>{

        act(()=>pageInfoSubscription.next(debugPage))
        expect(titleMock).toEqual({innerHTML: 'Debug'})
        expect(query.container).toMatchSnapshot()
    })

    it("When saving data, it will change the status, and launch onSave",()=>{
        act(()=>pageInfoSubscription.next(debugPage))
        const button = query.getByRole('button')
        act(()=>{button.dispatchEvent(new MouseEvent("click", { bubbles: true }))})
        expect(onSave).toHaveBeenCalledTimes(1)
        expect(onSave).toHaveBeenCalledWith({data:"other info", name: "Debug"})
        expect(query.container).toMatchSnapshot()
    })
})