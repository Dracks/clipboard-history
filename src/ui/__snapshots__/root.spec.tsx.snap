// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/root Should render the debug screen 1`] = `
<div>
  <div>
    Debug data
    <button>
      save
    </button>
  </div>
</div>
`;

exports[`ui/root When saving data, it will change the status, and launch onSave 1`] = `
<div>
  <div>
    other info
    <button>
      save
    </button>
  </div>
</div>
`;
